examples/io/src/util.cc

```diff
@@ -0,0 +1,87 @@
+/* Copyright 2023 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <fstream>
+
+#include "legateio.h"
+#include "util.h"
+
+#include "core/utilities/dispatch.h"
+#include "core/utilities/type_traits.h"
+
+namespace fs = std::filesystem;
+
+namespace legateio {
+
+namespace detail {
+
+struct write_fn {
+  template <legate::LegateTypeCode CODE, int32_t DIM>
+  void operator()(const legate::Store &store, const fs::path &path) {
+    using VAL = legate::legate_type_of<CODE>;
+
+    auto shape = store.shape<DIM>();
+    auto empty = shape.empty();
+    auto extents = empty ? legate::Point<DIM>::ZEROES()
+                         : shape.hi - shape.lo + legate::Point<DIM>::ONES();
+
+    int32_t dim = DIM;
+    int32_t code = store.code<int32_t>();
+
+    logger.print() << "Write a sub-array " << shape << " to " << path;
+
+    std::ofstream out(path, std::ios::binary | std::ios::out | std::ios::trunc);
+    // Each file for a chunk starts with the extents
+    for (int32_t idx = 0; idx < DIM; ++idx)
+      out.write(reinterpret_cast<const char *>(&extents[idx]),
+                sizeof(legate::coord_t));
+
+    if (empty)
+      return;
+    auto acc = store.read_accessor<VAL, DIM>();
+    // The iteration order here should be consistent with that in the reader
+    // task, otherwise the read data can be transposed.
+    for (legate::PointInRectIterator it(shape, false /*fortran_order*/);
```

Comment:
This does not compile for me and needs an explicit `<DIM>`.  A better option might be a helper function that can properly do template argument deduction. Even better - we should fix this in Legion so that the shape can return an iter with a common begin/end idiom and this can be written with `for (auto point : shape.iter(false))`